plutus-core/executables/plutus/AnyProgram/With.hs

```diff
@@ -52,34 +65,71 @@ withA s r = case s of
     SUnit       -> r
     STxSrcSpans -> r
 
-withFlatL :: forall s r. SLang s ->  (Flat (FromLang s) => r) -> r
-withFlatL s r = case s of
-    SPir sname sann -> withN @Flat sname $ withNT @Flat sname $ withA @Flat sann r
-    SPlc sname sann -> withN @Flat sname $ withNT @Flat sname $ withA @Flat sann r
-    SUplc sname sann -> withN @Flat sname $ withN @(ComposeC Flat UPLC.Binder) sname $
-                       withA @Flat sann r
-    SData -> error "Flat is not available for Data"
+withLangGeneral
+    :: forall constrTyName constrBinder constrName constrAnn constr s r.
+        ( constrTyName (FromNameTy 'Name)
+        , constrTyName (FromNameTy 'DeBruijn)
+        , constrTyName (FromNameTy 'NamedDeBruijn)
+        , constrBinder (UPLC.Binder UPLC.Name)
+        , constrBinder (UPLC.Binder UPLC.DeBruijn)
+        , constrBinder (UPLC.Binder UPLC.NamedDeBruijn)
+        , constrName (FromName 'Name)
+        , constrName (FromName 'DeBruijn)
+        , constrName (FromName 'NamedDeBruijn)
+        , constrAnn (FromAnn 'Unit)
+        , constrAnn (FromAnn 'TxSrcSpans)
+        , (forall tyname name ann. (constrTyName tyname, constrName name, constrAnn ann) =>
+            constr (PIR.Program tyname name UPLC.DefaultUni UPLC.DefaultFun ann))
+        , (forall tyname name ann. (constrTyName tyname, constrName name, constrAnn ann) =>
+            constr (PLC.Program tyname name UPLC.DefaultUni UPLC.DefaultFun ann))
+        , (forall name ann. (constrBinder (UPLC.Binder name), constrName name, constrAnn ann) =>
+            constr (UPLC.UnrestrictedProgram name UPLC.DefaultUni UPLC.DefaultFun ann))
+        )
+    => SLang s -> (constr (FromLang s) => r) -> r
+withLangGeneral s r = case s of
+    SPir sname sann ->
+          withNT @constrTyName sname
+        $ withN @constrName sname
+        $ withA @constrAnn sann r
+    SPlc sname sann ->
+          withNT @constrTyName sname
+        $ withN @constrName sname
+        $ withA @constrAnn sann r
+    SUplc sname sann ->
+          withN @(ComposeC constrBinder UPLC.Binder) sname
+        $ withN @constrName sname
+        $ withA @constrAnn sann r
+    SData -> error "not implemented yet"
 
-withShowL :: forall s r. SLang s ->  (Show (FromLang s) => r) -> r
-withShowL s r = case s of
-    SPir sname sann -> withN @Show sname $ withNT @Show sname $ withA @Show sann r
-    SPlc sname sann -> withN @Show sname $ withNT @Show sname $ withA @Show sann r
-    SUplc sname sann -> withN @Show sname $ withN @(ComposeC Show UPLC.Binder) sname $
-                       withA @Show sann r
-    SData -> r
+withLang
+    :: forall constr s r.
+        ( constr (FromNameTy 'Name)
+        , constr (FromNameTy 'DeBruijn)
+        , constr (FromNameTy 'NamedDeBruijn)
```

Comment:
We could give these three constraints a single name and reuse it here, in `withLangGeneral` and in `withNT`, plus do the same with `FromName`, `FromAnn` etc constraints, but I'm not sure if it buys us much and we can always do that later.